### ISSUE TITLE
New Minit 0.3, testing required

### ISSUE THREAD

**Issue by kasparsd** [kasparsd/minit#50]
I have rewritten most of the code to make it easier to maintain. 

It would be great if more people could test it before it's officially released. Here is the ZIP file https://github.com/kasparsd/minit/releases/tag/0.3-beta.1 or you can checkout the `feature/restructure`.
## New features
- All JS is placed in the footer by default. The only option to avoid that is by excluding the script from minit using the `minit-exclude-js` filter.
- The Minit `<script />` tag is now `async` by default.
- JS is processed by Minit only once when it's added in the footer, which should fix #38.
## Todo
- [ ] Exclude scripts with conditionals, etc. #42
- [x] Use plugin `version` for cache busting during updates.

**Comment by markoheijnen** [comment]
Will give it a try locally. I did stopped using it when I had some issues with it and since I moved to HTTP/2, I didn't needed it as much. I was wondering where you think this plugin will go with HTTP/2 around the corner.

**Comment by kasparsd** [comment]
Thanks @markoheijnen! Be sure to checkout the latest updates of the `feature/restructure` branch since I pushed some major changes very recently.

Re: HTTP/2 -- fewer requests will always be better. Secondly, there is no reason for serving multiple files when one is enough.

**Comment by bjornjohansen** [comment]
@kasparsd With HTTP/2, concatenating files is considered an anti-pattern. In fact, at best, you don’t hurt performance. There are several reasons why one would want to keep the files separate. One of them is cache busting: Instead of having to re-download the entire bundle, you just re-download the changed file.

I fear I just opened Pandora’s Box, and this will be a long discusion now :-)

**Comment by markoheijnen** [comment]
@bjornjohansen I guess I opened the box this case ;)

It would be nice to have a option to check for the protocol but also see the issue with page caching. So unsure yet how this could easily be integrated. I think that would mean integration by the developer itself.

**Comment by kasparsd** [comment]
@bjornjohansen Interesting. Is there any article you would recommend?

Re: cache invalidation -- Wouldn't it still need to make a HEAD request for each of the assets in order to check the last modified timestamp?

**Comment by kasparsd** [comment]
> also see the issue with page caching

@markoheijnen are you talking about the concatenated files being invalidated but still being referenced in cached HTML responses? Minit doesn't delete the expired cache files until you explicitly tell it to.

**Comment by bjornjohansen** [comment]
I’ll see what I can dig up, or maybe I can write one. I did a couple of presentations on it in April/May.

You will set expiry to `max` (or a long, long time into the future), and update the version number in the filename/querystring whenever the resource have changed. No `HEAD` request needed (and the browser won’t make them).

**Comment by kasparsd** [comment]
@bjornjohansen That would work!

**Comment by bjornjohansen** [comment]
Over HTTP/2 resources are downloaded in parallel (multiplexed over a single connection), do there is no connection overhead and resources are loaded as if they would be inlined. If you use prefetching or push, your external resources will in many cases load before your HTML document.

Since this plugin knows about all the enqueued JS files, I’ve many times thought of using it as a base for a JS prefetcher (sending prefetching-headers with the HTML).

**Comment by szepeviktor** [comment]
Works well.

**Comment by 1BJK903** [comment]
It does not work for me...

Fatal error: Call to undefined function wp_scripts() in /home/domains/website/public_html/wp-content/plugins/minit-0.3-beta.1/include/minit-js.php on line 12

PS: the 1.2 version does not work with Scribu's Front-End Editor plugin either. Same error: jQuery not defined.

**Comment by kasparsd** [comment]
Thanks for reporting @1BJK903! Which version of WP is that? That function was introduced in WordPress 4.2 https://core.trac.wordpress.org/browser/tags/4.4/src/wp-includes/functions.wp-scripts.php#L16

**Comment by 1BJK903** [comment]
No problem @kasparsd, you deserve this. It's a great plugin.
I am using Wordpress 4.1.. so that makes sense now :) Thanks!

**Comment by 1BJK903** [comment]
@kasparsd: I tested it with Wordpress 4.4..

Front-End Editor still not working + it breaks my fonts (using Google Fonts plugin) and styling of GDSR Rating plugin. So 3 essential plugins for me are not working with this beta version, while they are working with version 1.2

**Comment by szepeviktor** [comment]
GD Rating System 1.0.2.1 works for me.

**Comment by 1BJK903** [comment]
It's GD Star Rating 1.9.2.1 @szepeviktor. It's not that bad actually, cause it's just the styling.

**Comment by szepeviktor** [comment]
Do you have an URL for that plugin?

**Comment by szepeviktor** [comment]
@kasparsd Is it HTTP/2 server push?
https://www.webpagetest.org/result/160103_FD_RSY/1/details/
(click on a request and see `Content Download: ms`)

**Comment by 1BJK903** [comment]
@szepeviktor, it doesn't exist in the Wordpress repository anymore. Never mind, I am going to use the plugin you mentioned. It's newer.

**Comment by markoheijnen** [comment]
I'm seeing something weird. It does minit the dependencies but not the script itself. It's still stored in $this->queue but double. What I then see is the normal script tag and inside the minit-async-scripts script tag.

**Comment by szepeviktor** [comment]
@markoheijnen Are you on https://github.com/kasparsd/minit/commits/feature/restructure ?

**Comment by kasparsd** [comment]
@markoheijnen Can't replicate that. Is that script coming from a theme or a plugin? Can you post a dump of the HTML output somewhere?

**Comment by markoheijnen** [comment]
@szepeviktor: Yes, I am. HTML output is https://gist.github.com/markoheijnen/d43c62381f98f1c2631a. It's a local site of https://markoheijnen.com.

The script comes from the parent theme.

**Comment by markoheijnen** [comment]
It fails when checking if it's hosted locally.

**Comment by kasparsd** [comment]
@markoheijnen Yes, something isn't right here https://github.com/kasparsd/minit/blob/b03a5e5444012d1083fe1bf848561b51f12524aa/include/minit-assets.php#L208 -- probably something to do with using a custom `wp-contents` folder.

**Comment by markoheijnen** [comment]
Correct because $this->handler->base_url is always the wp folder. Currently playing around with it.

**Comment by szepeviktor** [comment]
Try using `WP_CONTENT_URL` instead.

**Comment by markoheijnen** [comment]
It isn't as simple to replace it with a different value. There needs a detection between WordPress folder and content folder.

**Comment by szepeviktor** [comment]
Yes, with custom `wp-content` location `file_exists( ABSPATH . $src )` will always return false in `minit()`.

**Comment by markoheijnen** [comment]
Created #60 that solves it for me. The double URLs when not a local URL is still something that need to be solved I think

**Comment by markoheijnen** [comment]
For me it does work perfectly with the additional fixes I just made. I did make it PHP5.3 only because of using preg_replace_callback with anonymous function. This for removing ../ in urls.
